Index: class/System/System.ComponentModel.Design/SelectionTypes.cs
===================================================================
--- class/System/System.ComponentModel.Design/SelectionTypes.cs	(revision 62189)
+++ class/System/System.ComponentModel.Design/SelectionTypes.cs	(working copy)
@@ -3,9 +3,10 @@
 //
 // Authors:
 //      Martin Willemoes Hansen (mwh@sysrq.dk)
+//      Ivan N. Zlatev (contact i-nZ.net)
 //
 // (C) 2003 Martin Willemoes Hansen
-//
+// (C) 2006 Ivan N. Zlatev
 
 //
 // Permission is hereby granted, free of charge, to any person obtaining
@@ -32,10 +33,34 @@
 
 namespace System.ComponentModel.Design
 {
+#if NET_2_0
 	[Flags]
+	[ComVisible(true)]
+	public enum SelectionTypes
+	{
+		Auto = 1,
+		[Obsolete ("This value has been deprecated. Use SelectionTypes.Auto instead.")]
+		Normal = 1,
+		Replace = 2,
+		[Obsolete ("This value has been deprecated. It is no longer supported.")]
+		MouseDown = 4,
+		[Obsolete ("This value has been deprecated. It is no longer supported.")]
+		MouseUp = 8,
+		[Obsolete ("This value has been deprecated. Use SelectionTypes.Primary instead.")]
+		Click = 16,
+		Primary = 16,
+		[Obsolete ("This value has been deprecated. It is no longer supported.")]
+		Valid = 31,
+		Toggle = 32,
+		Add = 64,
+		Remove = 128
+	}
+#else
+	
+	[Flags]
 	[Serializable]
 	[ComVisible(true)]
-        public enum SelectionTypes
+	public enum SelectionTypes
 	{
 		Click = 16,
 		MouseDown = 4,
@@ -44,4 +69,5 @@
 		Replace = 2,
 		Valid = 31
 	}
+#endif
 }
