--- DocTest-v1.cs	2006-12-07 15:33:43.000000000 -0500
+++ DocTest.cs	2006-12-07 15:37:07.000000000 -0500
@@ -1,6 +1,9 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using System.Reflection;
+
+[assembly: AssemblyVersion ("2.0.0.0")]
 
 namespace Mono.DocTest {
 	/// <summary>Possible colors</summary>
@@ -219,6 +222,9 @@
 		/// <remarks><c>M:Mono.DocTest.UseLists.UseHelper``3(Mono.DocTest.Generic.MyList{``0}.Helper{``1,``2})</c>.</remarks>
 		public void UseHelper<T,U,V> (Generic.MyList<T>.Helper<U,V> helper) {}
 	}
+
+	public class AddedType {
+	}
 }
 
 namespace Mono.DocTest.Generic {
@@ -234,6 +240,8 @@
 			return default(U);
 		}
 
+		public void AddedInVersion0_1_0_0 () {}
+
 		/// <param name="list">Insert description here</param>
 		/// <remarks><c>M:Mono.DocTest.GenericBase`1.op_Explicit(Mono.DocTest.GenericBase{`0})~`0</c></remarks>
 		/// <returns>The default value for <typeparamref name="U"/>.</returns>
