Index: Extras.mds
===================================================================
--- Extras.mds	(revision 83175)
+++ Extras.mds	(working copy)
@@ -21,9 +21,9 @@
       <Entry build="True" name="MonoDevelop.WebReferences" configuration="Debug" />
       <Entry build="True" name="Deployment" configuration="" />
       <Entry build="True" name="BooBinding" configuration="Debug" />
-      <Entry build="True" name="MonoQuery" configuration="Debug" />
       <Entry build="True" name="CSharpBinding" configuration="" />
       <Entry build="True" name="MonoDevelop.Gettext" configuration="Debug" />
+      <Entry build="True" name="MonoDevelop.Database" configuration="Debug" />
     </Configuration>
     <Configuration name="Release" ctype="CombineConfiguration">
       <Entry build="True" name="MonoDevelop.NUnit" configuration="Debug" />
@@ -46,9 +46,9 @@
       <Entry build="True" name="MonoDevelop.WebReferences" configuration="Release" />
       <Entry build="True" name="Deployment" configuration="" />
       <Entry build="True" name="BooBinding" configuration="Release" />
-      <Entry build="True" name="MonoQuery" configuration="Release" />
       <Entry build="True" name="CSharpBinding" configuration="" />
       <Entry build="True" name="MonoDevelop.Gettext" configuration="Release" />
+      <Entry build="True" name="MonoDevelop.Database" configuration="Release" />
     </Configuration>
   </Configurations>
   <StartMode startupentry="MonoDevelop.NUnit" single="True">
@@ -77,6 +77,7 @@
     <Execute type="None" entry="MonoQuery" />
     <Execute type="None" entry="CSharpBinding" />
     <Execute type="None" entry="MonoDevelop.Gettext" />
+    <Execute type="None" entry="MonoDevelop.Database" />
   </StartMode>
   <Entries>
     <Entry filename="./NUnit/MonoDevelop.NUnit.mdp" />
@@ -99,8 +100,8 @@
     <Entry filename="./MonoDevelop.WebReferences/MonoDevelop.WebReferences.mdp" />
     <Entry filename="./Deployment/Deployment.mds" />
     <Entry filename="./BooBinding/BooBinding.mds" />
-    <Entry filename="./MonoQuery/Mono.Data.Sql.mds" />
     <Entry filename="./CSharpBinding/CSharpBinding.mds" />
     <Entry filename="./MonoDevelop.Gettext/MonoDevelop.Gettext.mdp" />
+    <Entry filename="./MonoDevelop.Database/MonoDevelop.Database.mds" />
   </Entries>
 </Combine>
\ No newline at end of file
